woocommerce/woocommerce-ios
# Analytics Hub: Add Custom Range Support

part of #8200

# Why

This PR adds the logic required to support custom ranges in the Analytics Hub.
One decision we made is to compare the custom range against a range of the same length immediately before the start date of the chosen range.

**NOTE: As the date range UI is not integrated yet, this PR selects a fixed custom range temporarily.**

# How

- Creates `AnalyticsHubCustomRangeData` to format the custom range.

- Removes the required `init` from `AnalyticsHubTimeRangeData` Because protocols should indicate what a type has and not how it is created!

- Adds a `.custom` case in `AnalyticsHubTimeRangeSelection`

# Discussion

I would like us to decouple ranges supported by the business logic types vs ranges supported by the UI types. 

This PR by adding a case with an associated type, like `.custom:(Date?, Date?)` uncovers the limitations of coupling those and forces us to write code with a lot of conditionals that are not very clear for the reader.

[Issue](https://github.com/woocommerce/woocommerce-ios/issues/8340) to track 👆 

# Demo

https://user-images.githubusercontent.com/562080/206288557-739c5353-4bb6-4c8e-9877-15fe3d132bd5.mov

# Testing Steps

- Open the analytics Hub
- Select the custom range
- See that it loads data with a fixed custom range, currently from the start of the week until de end of today.
- See that the previous range ends before the start of the custom range and has the same length. 

---
